rclcpp/src/rclcpp/context.cpp

```diff
@@ -344,6 +344,7 @@ Context::shutdown(const std::string & reason)
 rclcpp::Context::OnShutdownCallback
 Context::on_shutdown(OnShutdownCallback callback)
 {
+  std::lock_guard<std::mutex> lock(on_shutdown_callbacks_mutex_);
```

Comment:
probably not only adding mutex here, but also all access to `on_shutdown_callbacks_`?

Reply:
I think this is not possible without changing the API as the reference is returned

Reply:
We should deprecate in the future those getters, returning a reference to an internal container isn't great.

I think that the mutex should also be taken [here](https://github.com/ros2/rclcpp/blob/6eb34a84e6009a73860bdf54418f34c4f81b2969/rclcpp/src/rclcpp/context.cpp#L318) before calling all the callbacks:

```cpp
  {
    std::lock_guard<std::mutex> lock(on_shutdown_callbacks_mutex_);
    for (const auto & callback : on_shutdown_callbacks_) {
      callback();
    }
  }
```

Reply:
> We should deprecate in the future those getters, returning a reference to an internal container isn't great.

In the general case, I don't agree with this, but in this particular case, I agree.

We could return a copy (since they should be trivially copyable) and lock that operation.

Or we could just drop the accessors. I think I added them just because having hidden state like that can bite you later if you need access to it and have to add an API to get it, or work around it by keeping track of them outside this class (duplicating the effort). So I'd lean towards returning a copy and protecting that with the mutex.